### PR TITLE
Micro hydras

### PR DESCRIPTION
**Pull request by Matuiss2**

### CODE REVIEW

File: actions/army_control.py [hunk 1]
```diff
@@ -136,6 +164,100 @@ def micro_zerglings(self, targets, unit):
...
+                if self.hydra_atk_range:
+                    our_range += 1
+                enemy_range = closest_threat.ground_range + closest_threat.radius
+                # For flying enemies,
+                if closest_threat.is_flying:
+                    our_range = unit.air_range + unit.radius
+                    if self.hydra_atk_range:
+                        our_range += 1
+                    # Hit and run if we can.
+                    if our_range > enemy_range and our_movespeed > closest_threat.movement_speed:
```
[the remark this answers is not in the document]
Comment (nikgrosland) [reply actions/army_control.py]: This check needs to be here. If it isn't, then we'll never hit_and_run against enemy air units. It also needs to be on line 204 or else we'll never hit_and_run against enemy ground units.

File: actions/army_control.py [hunk 2]
```diff
@@ -20,10 +24,33 @@
...
     ZERGLINGATTACKSPEED,
 )
-
+from sc2.position import Point2
 from .micro import Micro
 
 
+def find_pursuit_point(target, unit) -> Point2:
+    """Find a point towards the enemy unit"""
+    deltax = unit.position.x - target.position.x
```
[the remark this answers is not in the document]
Comment (nikgrosland) [reply actions/army_control.py]: I'll simplify this.

File: actions/army_control.py [hunk 3]
```diff
@@ -20,10 +24,33 @@
...
+    deltay = unit.position.y - target.position.y
+    return Point2((unit.position.x + ((deltax / 2) * -1), unit.position.y + ((deltay / 2) * -1)))
+
+
+def find_retreat_point(target, unit) -> Point2:
+    """Find a point away from the enemy unit"""
+    deltax = unit.position.x - target.position.x
```
[the remark this answers is not in the document]
Comment (nikgrosland) [reply actions/army_control.py]: I'll simplify this.

File: actions/army_control.py [hunk 4]
```diff
@@ -136,6 +164,100 @@ def micro_zerglings(self, targets, unit):
...
+            # Find the closest threat.
+            closest_threat = None
+            closest_threat_distance = math.inf
+            for threat in threats:
+                if threat.distance_to(unit) < closest_threat_distance and threat.ground_dps:
+                    closest_threat = threat
+                    closest_threat_distance = threat.distance_to(unit)
+            # If there's a close enemy that does damage,
+            if closest_threat:
+                our_range = unit.ground_range + unit.radius
```
[the remark this answers is not in the document]
Comment (nikgrosland) [reply actions/army_control.py]: You're right, both ranges are the same.

File: actions/army_control.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikgrosland) [reply actions/army_control.py]: I take it back, both ranges are the same so we don't need to differentiate between enemy air and ground units. I'll fix this.

[the remark this answers is not in the document]
Comment (Helfull) [reply actions/army_control.py]: even if you have to we do the same check, so just adjust all parameters before you do the check and you dont have to put it twice ;)

File: actions/army_control.py [hunk 5]
```diff
@@ -136,6 +164,100 @@ def micro_zerglings(self, targets, unit):
...
+        return self.attack_close_target(unit, targets)
+
+    def hit_and_run(self, target, unit):
+        """Attack when the unit can, run while it can't. We outrun the enemy."""
+        # Only do this when our range > enemy range, our movespeed > enemy movespeed, and enemy is targeting us.
+        local_controller = self.ai
+        action = local_controller.add_action
+        unit_is_air = unit.is_flying
+        target_is_air = target.is_flying
+        if target_is_air:
```
[the remark this answers is not in the document]
Comment (nikgrosland) [reply actions/army_control.py]: If you're talking about the target_is_air check, we can get rid of it entirely. It's from when I thought that a units air range could differ from its ground range. If we plan on extending this method to being used by more than just hydras, for instance calling it with a unit that can only attack ground or only attack air, then we will need to check if the enemy is flying and we can put that in a separate method.

**Comment by Matuiss2** [comment]
Ok, now its refactored, and include Nicnok's addressing of your review.

**Comment by Helfull** [comment]
@Matuiss2 getting the hang on the refactoring 👍

**Pull request merged**